repose-aggregator/components/services/opentracing-service/opentracing-service-impl/src/main/scala/org/openrepose/core/services/opentracing/TracerExtractor.scala

```diff
@@ -0,0 +1,105 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2015 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.core.services.opentracing
+
+import java.util
+import java.util.AbstractMap.SimpleImmutableEntry
+
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import io.opentracing.propagation.TextMap
+import org.openrepose.commons.utils.servlet.http.HttpServletRequestWrapper
+
+import scala.collection.mutable.ListBuffer
+
+class TracerExtractor() extends TextMap with LazyLogging {
+  var headers: Map[String, List[String]] = _
+
+  def this(httpServletRequestWrapper: HttpServletRequestWrapper) = {
+    this()
+
+    headers = servletHeadersToMultiMap(httpServletRequestWrapper)
+
+    def servletHeadersToMultiMap(httpServletRequest: HttpServletRequestWrapper): Map[String, List[String]] = {
+      val headersResult = scala.collection.mutable.HashMap.empty[String, List[String]]
+      if (httpServletRequest != null) {
+        val headerNamesIterator = httpServletRequest.getHeaderNames
+        while ( {
+          headerNamesIterator.hasMoreElements
+        }) {
+          val headerName: String = headerNamesIterator.nextElement
+          val valuesIterator = httpServletRequest.getHeaders(headerName)
+          var valuesList = ListBuffer[String]()
+          while ( {
+            valuesIterator.hasMoreElements
+          }) {
+            valuesList += valuesIterator.nextElement
+          }
+          headersResult += (headerName -> valuesList.toList)
+        }
+      }
+
+      headersResult.toMap
+
+    }
```

Comment:
You made this way more complex than it needs to be. Since you're taking a wrapped request anyway, use the convenience methods it provides. Also, leverage the power of Scala! 
```
  def servletHeadersToMultiMap(httpServletRequest: HttpServletRequestWrapper): Map[String, List[String]] = {
    Option(httpServletRequest).map(
      _.getHeaderNamesScala
        .map(headerName => headerName -> httpServletRequest.getHeadersScala(headerName))
        .toMap
    ).getOrElse(Map.empty)
  }
```

Reply:
And really, if you do that, you don't even need the nested/local method.